lib/receiver.ts

```diff
@@ -0,0 +1,289 @@
+        let onClose: Func<rhea.EventContext, void>;
+        let waitTimer: any;
+
+        const removeListeners = () => {
+          clearTimeout(waitTimer);
+          this._receiver.removeListener(ReceiverEvents.receiverError, onError);
+          this._receiver.removeListener(ReceiverEvents.receiverClose, onClose);
+        };
+
+        onClose = (context: rhea.EventContext) => {
+          removeListeners();
+          process.nextTick(() => {
+            log.receiver("[%s] Resolving the promise as the amqp receiver has been closed.",
+              this.connection.id);
+            resolve();
+          });
+        };
+
+        onError = (context: rhea.EventContext) => {
+          removeListeners();
+          log.error("[%s] Error occurred while closing amqp receiver. %O",
+            this.connection.id, context.session!.error);
+          reject(context.session!.error);
+        };
+
+        const actionAfterTimeout = () => {
+          removeListeners();
+          const msg: string = `Unable to close the amqp receiver ${this.name} due to operation timeout.`;
+          log.error("[%s] %s", this.connection.id, msg);
+          reject(new Error(msg));
+        };
+
+        this._receiver.once(ReceiverEvents.receiverClose, onClose);
+        this._receiver.once(ReceiverEvents.receiverError, onError);
+        waitTimer = setTimeout(actionAfterTimeout, defaultOperationTimeoutInSeconds * 1000);
+        this._receiver.close();
+      } else {
+        resolve();
+      }
+    });
+
+    return receiverClose.then(() => {
+      log.receiver("[%s] receiver has been closed, now closing it's session.", this.connection.id);
+      return this._session.close();
+    });
+  }
+
+  /**
+   * Registers the given handler for the specified event.
+   * @param {rhea.ReceiverEvents} event The event for which the handler needs to be registered.
+   * @param {rhea.OnAmqpEvent} handler The handler function that needs to be executed when the event
+   * occurs
+   * @returns {void} void.
+   */
+  registerHandler(event: ReceiverEvents, handler: rhea.OnAmqpEvent): void {
+    this._receiver.on(event, handler);
+  }
+
+  /**
+   * Removes the given handler for the specified event.
+   * @param {rhea.ReceiverEvents} event The event for which the handler needs to be removed.
+   * @param {rhea.OnAmqpEvent} handler The handler function that needs to be removed.
+   * @returns {void} void.
+   */
+  removeHandler(event: ReceiverEvents, handler: rhea.OnAmqpEvent): void {
+    this._receiver.removeListener(event, handler);
+  }
+
+  /**
+   * Registers the given handler for the specified event on the session object.
+   * @param {rhea.SessionEvents} event The event for which the handler needs to be registered.
+   * @param {rhea.OnAmqpEvent} handler The handler function that needs to be executed when the event
+   * occurs
+   * @returns {void} void.
+   */
+  registerSessionHandler(event: rhea.SessionEvents, handler: rhea.OnAmqpEvent): void {
```

Comment:
continuing with the event emitter comment, i'd be inclined to just let people access the session itself from the link and then register the listeners they want using standard event emitter patterns. so this:

```ts
myReceiver.registerSessionHandler('someEvent', () => {});
```

becomes:

```ts
myReceiver.session.on('someEvent', () => {});
```